scripts/shtf.py

```diff
@@ -0,0 +1,629 @@
+
+
+class MessageType(TypedDict):
+    text: str
+    message_link: NotRequired[str]
+
+
+TokensConfig = dict[str, list[tuple[str, str]]]
+
+
+class ConfigType(TypedDict):
+    agent_address: str
+    deployment_dir: Path
+    notification_system: str
+    rpcs: dict[int, tuple[str, BlockNumber]]
+    explorers: dict[int, str]
+    notification: dict[str, dict[str, str]]
+    tokens: TokensConfig
+
+
+class TransferType(TypedDict):
+    created: NotRequired[RequestCreated]
+    filled: NotRequired[RequestFilled]
+    claimed: NotRequired[dict[int, list[ClaimMade]]]
+    withdrawn: NotRequired[DepositWithdrawn]
+
+
+transfersType = DefaultDict[str, TransferType]
+
+ChainToEvent = dict[ChainId, list[beamer.events.Event]]
+
+
+class MetaType(TypedDict):
+    request_id: str
+    message_type: str
+    message_link: str
+
+
+class NotificationType(TypedDict):
+    meta: MetaType
+    body: str
+
+
+@dataclass
+class Stats:
+    fills: int = 0
+    requests: int = 0
+    withdraws: int = 0
+    volume: int = 0
+
+
+@dataclass
+class TransferStats:
+    volume: int = 0
+    fills: int = 0
+    requests: int = 0
+    withdrawals: int = 0
+    claims: int = 0
+    expired_requests: int = 0
+    expired_requests_volume: int = 0
+
+
+class NotificationState:
+    def __init__(self) -> None:
+        with open(os.path.join(os.path.dirname(__file__), "breadcrumbs.json"), "r") as f:
+            self.notified_state = json.load(f)
+
+    def is_set(self, request_id: str, notified_on: str) -> bool:
+        breadcrumb = self.notified_state.get(str(request_id))
+        if breadcrumb is None:
+            return False
+        else:
+            return breadcrumb.get(notified_on, False)
+
+    def update(self, request_id: str, notified_on: str) -> None:
+        breadcrumb = self.notified_state.get(str(request_id))
+        if breadcrumb is None:
+            self.notified_state[str(request_id)] = dict()
+            self.notified_state[str(request_id)][notified_on] = True
+        else:
+            breadcrumb[notified_on] = True
+
+    def persist(self) -> None:
+        with open(os.path.join(os.path.dirname(__file__), "breadcrumbs.json"), "w") as f:
+            json.dump(self.notified_state, f)
+
+
+@dataclass
+class Context:
+    stats: TransferStats = field(default_factory=TransferStats)
+    messages: list = field(default_factory=list)
+    agent_address: str = ""
+
+    notification_state: NotificationState = field(default_factory=NotificationState)
+
+    def add_message(self, message: NotificationType) -> None:
+        self.messages.append(message)
+
+
+class Notify:
+    def __init__(self, system: str, config: dict[str, dict[str, str]]) -> None:
+        self.system = system
+        self.config = config
+
+    def send(self, message: MessageType, callback: Optional[Callable] = None) -> None:
+        global notifications_sent
+        if notifications_sent > 0 and notifications_sent % 5 == 0:
```

Comment:
just looked at notifications_sent and it is only used in Notify class why we are using global here ?

Reply:
I've moved it to the notify class now.